File: src/lab0/src/baseline.py
from pathlib import Path
import time


def read_matrix(path: Path) -> list[list[float]]:
    with open(path) as f:
        next(f)  # Skip the first line
        return [list(map(float, line.split())) for line in f]


def write_matrix(path: Path, matrix: list[list[float]]):
    with open(path, "w") as f:
        for row in matrix:
            f.write(" ".join(map(str, row)) + "\n")


def matrix_multiply(A, B):
    n = len(A)
    m = len(A[0])
    p = len(B[0])
    C = [[0.0] * p for _ in range(n)]
    for i in range(n):
        for j in range(p):
            for k in range(m):
                C[i][j] += A[i][k] * B[k][j]
    return C


def main():
    A = read_matrix(Path("data/A.txt"))
    B = read_matrix(Path("data/B.txt"))
    start = time.perf_counter()
    C = matrix_multiply(A, B)
    end = time.perf_counter()
    write_matrix(Path("data/output.txt"), C)
    print(f"Time taken: {(end - start) * 1000:.2f} ms")


if __name__ == "__main__":
    main()

File: src/lab0/src/baseline.cpp
#include <chrono>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace std;
vector<vector<float>> read_matrix(const string& filename) {
	int n = 0, m = 0;
	ifstream fin(filename);
	fin >> n >> m;
	vector<vector<float>> matrix(n, vector<float>(m));
	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < m; ++j) {
			fin >> matrix[i][j];
		}
	}
	fin.close();
	return matrix;
}

vector<vector<float>> matrix_multiply(const vector<vector<float>>& A, const vector<vector<float>>& B) {
	vector<vector<float>> C(A.size(), vector<float>(B[0].size()));
	for (int i = 0; i < (int)A.size(); ++i) {
		for (int j = 0; j < (int)B[0].size(); ++j) {
			for (int k = 0; k < (int)A[0].size(); ++k) {
				C[i][j] += A[i][k] * B[k][j];
			}
		}
	}
	return C;
}

int main() {
	auto A = read_matrix("data/A.txt");
	auto B = read_matrix("data/B.txt");
	auto start = chrono::steady_clock::now();
	auto C = matrix_multiply(A, B);
	auto end = chrono::steady_clock::now();
	ofstream fout("data/C.txt");
	for (int i = 0; i < (int)C.size(); ++i) {
		for (int j = 0; j < (int)C[0].size(); ++j) {
			fout << C[i][j] << ' ';
		}
		fout << '\n';
	}
	fout.close();
	auto duration = chrono::duration_cast<chrono::milliseconds>(end - start).count();
	cout << "Time taken: " << duration << " ms" << endl;
	return 0;
}